docs/adr/002adr-types.md

```diff
@@ -0,0 +1,326 @@
+* `F` is a set of functions from a tuple of integers to an integer:
+    `F <: "Set(<<Int, Int>> -> Int)"`.
+* `Foo` is an operator of an integer and of a string that returns an integer:
+    `Foo <: "(Int, Str) => Int"`.
+* `Bar` is a higher-order operator that takes an operator that takes
+    an integer and a string and returns an integer, and returns a Boolean:
+    `Bar <: "((Int, Str) => Int) => Bool"`.
+* `Baz` is a polymorphic operator that takes two arguments of the same type
+    and returns a value of the type equal to the types of its arguments:
+    `Baz <: "(a, a) => a"`.
+* `Proc` and `Faulty` are sets of the same type:
+    `Proc <: "Set(PID)"` and `Faulty <: "Set(PID)"`.
+
+### 1.2. Discussion    
+
+Our type grammar presents a minimal type system that, in our understanding,
+captures all interesting cases that occur in practice. Obviously, this type
+system considers ill-typed some perfectly legal TLA+ values. For instance, we
+cannot assign a reasonable type to `{1, TRUE}`. However, we can assign a
+reasonable type to `{[type |-> "1a", bal |-> 1], [type |-> "2a", bal |-> 2, val
+|-> 3]}`, a pattern that often occurs in practice, e.g., see
+[Paxos](https://github.com/tlaplus/Examples/blob/master/specifications/Paxos/Paxos.tla).
+The type of that set will be `Set([type: Str, bal: Int, val: Int])`, which is
+probably not what you expected, but it is the best type we can actually compute
+without having algebraic datatypes in TLA+. It also reminds the user that one
+better tests the field `type` carefully.
+
+Type System 1 is also very much in line with the [type system by Stephan Merz
+and Hernan
+Vanzetto](https://dblp.org/search?q=Automatic+Verification+of+%7BTLA%7D+%2B+Proof+Obligations+with+%7BSMT%7D+Solvers),
+which is used internally by
+[TLAPS](https://tla.msr-inria.inria.fr/tlaps/content/Home.html) when translating
+proof obligations in SMT.  We introduce types for user-defined operators, on top
+of their types for TLA+ expressions that do not contain user-defined operators.
+
+We expect that this type system will evolve in the future. That is why we call
+it __Type System 1__. Feel free to suggest __Type System 2__ :-)
+
+## 2. How to write type annotations (as a user)
+
+We define the Apalache module `Typing.tla` that contains definitions of two
+infix operators:
+
+```tla
+---- MODULE Typing ----
+ex <: tp == TRUE
+tp ## ex == ex
+=======================
+```
+
+The operator `ex <: tp` is a type assumption. It states that `ex` has the type
+whose supertype is `tp` (the records in `tp` may contain additional fields).
+This operator always returns `TRUE`.  The operator `tp ## ex` annotates an
+expression `ex` with a type `tp`.  This operator return `ex` itself, that is, it
+performs type erasure (for compatibility with other TLA+ tools).  As these
+operators return values of different types, we decided to use different
+graphical notation for them. Initially, we thought of writing `<:` and `:>`,
```

Comment:
fiwiw, when I suggested `:>` it was in a context where `<:` was defined as

```
value <: type == value
```

enabling *inline* ascription via erasure. In this context `<:` and `:>` have a very natural symmetry. 
This is also our current practice, iiuc. 

Your proposed redefinition of the `<:` operator does break this symmetry, but it would be possible to keep it by pursuing either of these options:

1. Add an operator like `AssumeType(ascription) == TRUE" which could be used like `AssumeType(value <: Type)`
2. Use `##` for the type assumptions and leave the symmetrical pair `<:` and `:>` for inline type ascription.

Regardless of the particular symbols we use, only having these two operators means it's not possible to do inline type ascriptions in the now widely expected order of `LET foo : type == bar IN`, in which the type follows after the value. This might be nice to support one way or another.

Reply:
I thought of `AssumeType(ascription) == TRUE` too, but that is getting too verbose. Plus to that, I would like to avoid `e <: tp` returning `e`, as it would suggest those type annotations could be placed everywhere. That is actually the problem of the current approach: There are several places where one has to care about annotations. If we only annotate names (variables and operators), this makes annotations well isolated.

Reply:
Regarding your last comment, why would not `LET foo == type ## bar IN` work?

Reply:
We can also introduce the symmetrical version of `<:`. However, its use is going to be isolated in the current approach, so what would be the benefit?

Reply:
Just to preface: I'm happy to explain my reasoning and thinking here, but I'm not trying to block the ADR or anything.

> I would like to avoid e <: tp returning e, as it would suggest those type annotations could be placed everywhere. That is actually the problem of the current approach: There are several places where one has to care about annotations. If we only annotate names (variables and operators), this makes annotations well isolated.

Ah,  I wasn't aware of that. However, wouldn't `##`(or  `:>`) have the same problem? It seems like any inline type ascription operator defined via erasure will invite use anywhere, regardless of order of arguments. Or are we planning on adding something that will throw an error if the operator is used outside of isolated situations?

> Regarding your last comment, why would not `LET foo == type ## bar IN` work?

It would work, it's just that `LET foo <: type = bar IN` looks almost exactly like 

```sml
let foo : type = bar
```

which is familiar syntax for anyone who has used languages like Rust, Scala, Haskell, OCaml, TypeScript, Swift or anything else that follows ML. On the other hand, I've never seen inline type annotation syntax that looks like `let foo == type ## value`. So this is an appeal to familiarity, and working with icons and idioms that are likely to be familiar to people already. My suggestion for `:>` (or we could use `##`, tho you lose the hint provided by `:`, and IMO the iconic quality of notation is really quite important) was for the special case of a workaround for annotating operators. That said, if this order makes good sense to you and feels natural, it's probably not very important whether the type is to the left or right of the equality sign!

One thing we might consider if we go the two-operator route: perhaps we could use `:>` for inline ascriptions and `##` (or maybe just `TypeOf(value, Type)` for "type assumptions". My rational is that `LET foo == type :> bar` is already unfamiliar in putting the type on the right of the `==`, but at least it carries the hint of the `:`. Whereas 

```
/\ value1 ## type
/\ value2 ## type
```

Is at least familiar in that it goes `<identifier> op <type>` and 

```
/\ TypeOf(value1, type)
/\ TypeOf(value2, type)
```

is just fully self documenting.

Reply:
Good points! I would also love to see `LET foo: type = bar IN`. However, that is very much outside of our control. The canonical syntax for TLA definitions is `LET foo == bar...` (or without LET for the top-level definitions). Maybe one day Leslie will like the idea of type annotations so much that we will be allowed to extend the standard syntax :-)

I like your idea of keeping `:>` for the operator definitions and introducing `TypeOf` for the type assumptions. Let's even call this operator `AssumeType`, which is exactly its semantics.

My current understanding is that we should try to limit `:>` to operator annotations. It is easy to work around this constraint by introducing a LET-definition. The bonus is however huge, as the users will have a very clear rule of what should be annotated (constants, variables, and operator definitions) and how they should be annotated. This is close to what Scala does. Though in Scala one can write `.asInstance` :-)